### PR TITLE
Preserve values

### PR DESCRIPTION
**Pull request by 15r10nk**
Fix incorrect values and preserve correct values in the same structure.

This will allow to combine inline-snapshot with libraries like dirty-equals.

Features:

* compares lists, dicts and tuples recursively 
* changes only the code which has a different value (with fix)
* updates only the parts which have a different representation (with update)
* lists and tuples are aligned (like in a source diff) and the algorithm tries to maximizes the number of matching values.

### CODE REVIEW

**Comment by 15r10nk** [comment]
hi @alexmojaki,

This is the upcoming feature which supports the combination of dirty-equals with inline-snapshots. You could help me with testing this branch for your use cases.

**Comment by alexmojaki** [comment]
Thanks. I just tried it and found a problem: the order of dict items is not preserved. The code gets updated to reflect the order of the actual value, whereas I'd prefer it used the order in the existing snapshot code where possible, at least for `fix` (as opposed to `update`).

**Comment by 15r10nk** [comment]
the order in the following example is preserved:

``` python
from inline_snapshot import snapshot


def test_function():
    assert {2: 3, 1: 5} == snapshot({1: 5, 2: 6})
```

`pytest --inline-snapshot=fix`

``` python
from inline_snapshot import snapshot


def test_function():
    assert {2: 3, 1: 5} == snapshot({1: 5, 2: 3})
```

I think I implemented the behavior which you described. Changing the order with `update` is currently not implemented.

Can you give me an example where inline-snapshot does not what you want it to do?

**Comment by alexmojaki** [comment]
OK, I think the problem is just that I should have uninstalled inline-snapshot before running `pip install git+https://github.com/15r10nk/inline-snapshot.git@preserve-values` .

File: inline_snapshot/_change.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+    is_start = True
+    elements = 0
+
+    for index, entry in enumerate(parent_elements):
+        if index in to_insert:
+            new_code += to_insert[index]
+            print("insert", entry, new_code)
+        if entry is None:
+            deleted = True
+            print("delete1", entry)
```
Comment (alexmojaki) [inline inline_snapshot/_change.py:106]: I noticed these prints while testing just now

Comment (15r10nk) [reply inline_snapshot/_change.py:106]: thank you. they are long gone in my local version.

Comment (15r10nk) [reply inline_snapshot/_change.py]: I pushed the new version. There should be no difference for you, except the print statements which are gone now.

**Comment by 15r10nk** [comment]
hi @alexmojaki,

I'm pretty much done with the implementation (documentation still missing). I changed almost the complete implementation and you could do me a favor if you could use the last version of this branch to test it.

File: docs/eq_snapshot.md [hunk 2]
```diff
@@ -1,7 +1,7 @@
 ## General
 
 A snapshot can be compared against any value with `==`.
-The value gets recorded if the snapshot is undefined (`snapshot()`)
+The value can be recorded with `--inline-snapshot=create` if the snapshot is empty.
```
Comment (alexmojaki) [inline docs/eq_snapshot.md:4]: FYI, as described in https://github.com/15r10nk/inline-snapshot/issues/57#issuecomment-2018930088, I use `--inline-snapshot=fix` very liberally so that I very rarely have to type it. This means I also never use `create`, I just type `snapshot(0)` (because `0` is easiest to type inside `()`) and then `fix` the `0`. I don't know what the advantage is of requiring `create` for `snapshot()`, why not just make `fix/update` work there as well?

File: docs/eq_snapshot.md [hunk 3]
```diff
@@ -19,9 +19,154 @@ Example:
...
+=== "--inline-snapshot=fix"
+    <!-- inline-snapshot: fix -->
+    ```python
+    def test_something():
+        assert 2 + 3 == snapshot(5)
+    ```
+
+The value can later be changed with `--inline-snapshot=fix` if the value the snapshot is compared with has changed.
+
+## dirty-equals
```
Comment (alexmojaki) [inline docs/eq_snapshot.md:38]: I think it makes sense to mention dirty-equals because it's a good use case for this feature and they make a great combination. But I would suggest first explaining the general feature so that people don't have to learn about dirty-equals. Here's an example of how it can make tests more readable:

```python
def summary(*values):
    return {"total": sum(values)}


def test_summary():
    assert summary(1, 2, 3) == snapshot({"total": 1 + 2 + 3})  # clearer than {"total": 6}
```

Here's how you can present your example without dirty-equals:

```python
def test_function():
    result = get_data()
    assert result == snapshot(
        {"date": result["date"], "payload": "some data"}
    )
    assert isinstance(result["date"], datetime.datetime)
```

File: docs/eq_snapshot.md [hunk 4]
```diff
@@ -19,9 +19,154 @@ Example:
...
+        return {
+            "date": datetime.datetime.now(),
+            "payload": "some data",
+        }
+
+
+    def test_function():
+        assert get_data() == snapshot(
+            {
+                "date": IsDatetime(),
```
Comment (alexmojaki) [inline docs/eq_snapshot.md:109]:
```suggestion
                "date": IsNow(),
```

https://dirty-equals.helpmanual.io/latest/types/datetime/#dirty_equals.IsNow

**Review by alexmojaki — COMMENTED** [review]
I've been using this branch without problems, including in snapshots that contain non-dirty-equals variable expressions. I'm very excited for this to be released and to tell my coworkers about it!

File: docs/eq_snapshot.md (hunk 2 above)
Comment (15r10nk) [reply docs/eq_snapshot.md:4]: the idea is that create will only create new snapshots which are currently empty and **not change existing one**.

This allows to add new `snapshot()` and create the empty one without having to fear to accidentally changing existing ones.

This is only an option for the user. You can also use `--inline-snapshot=create,fix` to do both at the same time.

I will try to explain this better when I redo some parts of the documentation.

File: docs/eq_snapshot.md (hunk 4 above)
Comment (15r10nk) [reply docs/eq_snapshot.md:107]: General a good idea, but I can not use `now` in my documentation, because this would give me always a different value for the first example, which would break my tests ( I execute the code snippets in the documentation in my tests)

File: docs/eq_snapshot.md (hunk 2 above)
Comment (alexmojaki) [reply docs/eq_snapshot.md:4]: But `fix` doesn't work for empty snapshots, right?

Comment (15r10nk) [reply docs/eq_snapshot.md:4]: No, this is the reason why flags can be combined like `--inline-snapshot=create,fix`.
Every change is assigned to one flag. And the flags allow you to select which changes you want to apply.

Comment (alexmojaki) [reply docs/eq_snapshot.md:4]: What's the advantage of `fix` not working with `snapshot()`? Why force the user to use `create`?

File: docs/eq_snapshot.md (hunk 3 above)
Comment (alexmojaki) [reply docs/eq_snapshot.md:38]: > dirty-equals are handled special when `--inline-snapshot=update` is used (they are preserved). This is not the case for your examples.

What's the motivation for this special case?

> I intentional decided to limit the documentation/implementation to dirty-equals, because I did not know how a general use case might look like or what the problem could be that the user might try to solve.

What about the examples above?

> I would prefer to implement features when I understand the usage scenario the user might have. But I have in general nothing against it, just in a later version maybe.

What's the usage scenario for `update` and how does it relate to dirty-equals vs other expressions? I haven't had a need for `update` myself.

You make it sound like this has only been implemented for dirty-equals, but doesn't `fix` work equally well for all kinds of expressions?

I think it's important for users to know about the general feature. It's useful information and not everyone will want to use dirty-equals. As I've said, some of my own use cases involve non-dirty-equals expressions.

File: docs/eq_snapshot.md (hunk 2 above)
Comment (15r10nk) [reply docs/eq_snapshot.md:4]: there are currently 4 flags which have no overlapping (create,fix,trim,update)

https://15r10nk.github.io/inline-snapshot/pytest/

(I think the flags should be called categories, this would fit better)

The thing is that you could ask the same question for update. should fix imply update?

This is actually sometime the case in the current version and it is not easy to explain when fix will also update other snapshots. This should be gone with the new implementation.

I wanted to divide this reasons for snapshot changes to give the user the control above what he wants to change.

Long story short, I could make fix work for `snapshot()` but this would mix these categories/flags together and the benefit would be that the user does not have to type some characters (which are probably in already in your history). I wanted to separate these categories and chose the current implementation.

File: docs/eq_snapshot.md (hunk 3 above)
Comment (15r10nk) [reply docs/eq_snapshot.md:38]: The idea behind update is that the `__repr__()` implementation of some object might change (defaults are excluded from the repr, the class was renamed in a backward compatible way and is available with the new and old name, or you updated to a newer inline-snapshot version which creates multiline strings).

update changes the cases where old_value==new_value but repr(new_value) is different from the current source.

I have currently no way to decide if the user changed the code (like in your example) or if __repr__() was changed. 
This is the reason why I need a way to detect if update should change the code (the current implementation checks it with `isinstance(old_value,DirtyEquals)` ).

> What about the examples above?

They did not looked like to me like examples which you have created to have some examples. Sorry if this is not the case.

> As I've said, some of my own use cases involve non-dirty-equals expressions.

Thank you for mentioning this again. I think you mean some `[...]*10` cases right?
My best idea currently is to wrap these changes in something like `no_update([...]*10)` or to accept the fact that an update would revert your changes.

Comment (15r10nk) [reply docs/eq_snapshot.md:38]: hi @alexmojaki, I created an example which shows how how this could be implemented with dirty-equals.

``` python
from inline_snapshot import snapshot
from dirty_equals import DirtyEquals


class NoSnapshotUpdate(DirtyEquals):
    def __init__(self, value):
        self.__value = value
        super().__init__(value=value)

    def equals(self, other):
        return other == self.__value


def test_blub():
    assert {
        1: 5,
        2: 2,
    } == snapshot(
        {
            1: 1,
            2: NoSnapshotUpdate(1 + 1),
        }
    )
```
The 1+1 will not be updated, even if you run it with update

``` bash
pytest --inline-snapshot=fix,update
```

you can either use something like this, or never use `update`. I see no other way to combine both worlds.

I would also like to stick with the current dirty-equals support. It could become to complicated to support generic solution which might be used in a variety of ways.

But also thank you that you convinced me to support dirty-equals. I'm really happy with this idea :+1: .

Comment (alexmojaki) [reply docs/eq_snapshot.md:38]: I don't use `update` and I don't think its exact behaviour is that important, I was just surprised that dirty-equals was being treated specially.

I still think it's very useful for readers to know that `fix` preserves other expressions, but that's just my opinion and definitely not something that needs to block this PR.

**Comment by 15r10nk** [comment]
@alexmojaki, 0.8.0 has been released. Thank you very much for your feedback. I will clarify the different flags soon when I revise some parts of the documentation.
